### PR TITLE
added new resource  bq/routine

### PR DESCRIPTION
**Pull request by edwardmedia**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/4493

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_bigquery_routine`
```

### CODE REVIEW

File: products/bigquery/api.yaml [hunk 1]
```diff
@@ -1433,3 +1433,138 @@ objects:
...
+        required: true
+        properties:
+          - !ruby/object:Api::Type::String
+            name: 'datasetId'
+            description: The ID of the dataset containing this routine
+            required: true
+          - !ruby/object:Api::Type::String
+            name: 'projectId'
+            description: The ID of the project containing this routine
+            #  required: true
```
[the remark this answers is not in the document]
Comment (ScottSuarez) [reply products/bigquery/api.yaml:1460]: I see you are inherited project if not defined. Suggest explaining here or removing commented out code

**Review by ScottSuarez — COMMENTED** [review]
Made some comments for revision. Adding @rileykarson

File: products/bigquery/api.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (edwardmedia) [reply products/bigquery/api.yaml]: Yes, it is required from GCP but the provider will use the default value if users don't provide. This is the behavior across the board, would consider not adding the comment specific for this resource ( the comment pointed out will be removed)

File: products/bigquery/api.yaml [hunk 2]
```diff
@@ -1433,3 +1433,138 @@ objects:
...
+                Required unless argumentKind = ANY_TYPE. ~>**NOTE**: Because this field expects
+                a JSON string, any changes to the string will create a diff, even if the JSON
+                itself hasn't changed. If the API returns a different value for the same schema,
+                e.g. it switched the order of values or replaced STRUCT field type with RECORD
+                field type, we currently cannot suppress the recurring diff this causes. As a
+                workaround, we recommend using the schema as returned by the API.
+      - !ruby/object:Api::Type::String
+        name: 'returnType'
+        description: |
+          Optional A JSON schema for the return type . if language = "SQL"; required otherwise.
```
[the remark this answers is not in the document]
Comment (edwardmedia) [reply products/bigquery/api.yaml]: Rephrased by moving ` A JSON schema for the return type` in the front.

File: products/bigquery/api.yaml [hunk 3]
```diff
@@ -1433,3 +1433,138 @@ objects:
...
+          cannot suppress the recurring diff this causes. As a workaround, we recommend using
+          the schema as returned by the API.
+      - !ruby/object:Api::Type::Array
+        name: 'importedLibraries'
+        description: |
+          Optional. If language = "JAVASCRIPT", this field stores the path of the
+          imported JAVASCRIPT libraries.
+        item_type: Api::Type::String
+      - !ruby/object:Api::Type::String
+        name: 'definitionBody'
```
[the remark this answers is not in the document]
Comment (edwardmedia) [reply products/bigquery/api.yaml:1550]: `required: true` is added

**Review by ScottSuarez — APPROVED** [review]
lgtm. @rileykarson could you give it a pass as well whenever you have time to spare?

File: products/bigquery/api.yaml [hunk 4]
```diff
@@ -1433,3 +1433,139 @@ objects:
...
+        name: routineReference
+        description: Reference describing the ID of this routine
+        required: true
+        properties:
+          - !ruby/object:Api::Type::String
+            name: 'datasetId'
+            description: The ID of the dataset containing this routine
+            required: true
+          - !ruby/object:Api::Type::String
+            name: 'projectId'
```
Comment (rileykarson) [inline products/bigquery/api.yaml:1458]: nit: this should be `required: true` as well, even though Terraform isn't using it.

File: products/bigquery/api.yaml [hunk 5]
```diff
@@ -1433,3 +1433,139 @@ objects:
...
+            name: 'routineId'
+            description: The ID of the the routine. The ID must contain only
+              letters (a-z, A-Z), numbers (0-9), or underscores (_).
+              The maximum length is 256 characters.
+            required: true
+      - !ruby/object:Api::Type::Enum
+        name: 'routineType'
+        description: The type of routine.
+        values:
+          - :ROUTINE_TYPE_UNSPECIFIED
```
Comment (rileykarson) [inline products/bigquery/api.yaml:1470]:
```suggestion
```

`_UNSPECIFIED` enum values are usually an artifact of some internal rules that protobuf enums need to have their first value be `_UNSPECIFIED` or some kind of default. They generally aren't configurable, so we can probably remove this. If you've found it is configurable, feel free to ignore this!

File: products/bigquery/api.yaml [hunk 6]
```diff
@@ -1433,3 +1433,139 @@ objects:
...
+            name: 'projectId'
+            description: The ID of the project containing this routine
+          - !ruby/object:Api::Type::String
+            name: 'routineId'
+            description: The ID of the the routine. The ID must contain only
+              letters (a-z, A-Z), numbers (0-9), or underscores (_).
+              The maximum length is 256 characters.
+            required: true
+      - !ruby/object:Api::Type::Enum
+        name: 'routineType'
```
Comment (rileykarson) [inline products/bigquery/api.yaml:1467]: I would be fairly surprised if this field was updatable- can you test it?

File: products/bigquery/api.yaml [hunk 7]
```diff
@@ -1433,3 +1433,139 @@ objects:
...
+          The time when this routine was created, in milliseconds since the
+          epoch.
+      - !ruby/object:Api::Type::Integer
+        name: 'lastModifiedTime'
+        output: true
+        description: |
+          The time when this routine was modified, in milliseconds since the
+          epoch.
+      - !ruby/object:Api::Type::Enum
+        name: 'language'
```
Comment (rileykarson) [inline products/bigquery/api.yaml:1486]: Same comments as above- we can likely clear the UNSPECIFIED entry and I'd be surprised if we can update the field, so let's prove it in a test.

File: products/bigquery/api.yaml [hunk 8]
```diff
@@ -1433,3 +1433,139 @@ objects:
...
+        description: Input/output argument of a function or a stored procedure.
+        item_type: !ruby/object:Api::Type::NestedObject
+          properties:
+            - !ruby/object:Api::Type::String
+              name: 'name'
+              description: |
+                The name of this argument. Can be absent for function return argument.
+            - !ruby/object:Api::Type::Enum
+              name: 'argumentKind'
+              description: Defaults to FIXED_TYPE.
```
Comment (rileykarson) [inline products/bigquery/api.yaml:1504]: Can we add this as the default explicitly?

File: products/bigquery/api.yaml [hunk 9]
```diff
@@ -1433,3 +1433,139 @@ objects:
...
+              values:
+                - :ARGUMENT_KIND_UNSPECIFIED
```
Comment (rileykarson) [inline products/bigquery/api.yaml:1506]: (same drill as above re: unspecified for the rest of these enums)

File: products/bigquery/api.yaml (hunk 7 above)
Comment (edwardmedia) [reply products/bigquery/api.yaml:1486]: Cannot replace routine of type PROCEDURE with a different type SCALAR FUNCTION.

File: products/bigquery/api.yaml (hunk 6 above)
Comment (edwardmedia) [reply products/bigquery/api.yaml:1467]: "Cannot replace routine of type PROCEDURE with a different type SCALAR FUNCTION"

File: products/bigquery/api.yaml (hunk 7 above)
Comment (edwardmedia) [reply products/bigquery/api.yaml:1486]: Language can be updated. Verified in a test

File: products/bigquery/api.yaml (hunk 9 above)
Comment (edwardmedia) [reply products/bigquery/api.yaml]: removed

File: products/bigquery/api.yaml (hunk 8 above)
Comment (edwardmedia) [reply products/bigquery/api.yaml:1504]: added

File: products/bigquery/api.yaml (hunk 5 above)
Comment (edwardmedia) [reply products/bigquery/api.yaml]: removed

File: products/bigquery/api.yaml (hunk 4 above)
Comment (edwardmedia) [reply products/bigquery/api.yaml:1458]: added

**Review by rileykarson — APPROVED** [review]
Thanks for getting working Javascript examples @edwardmedia!

**Pull request merged**